commons-mongo/jvm/src/main/scala-2.13/com/avsystem/commons/mongo/typed/MongoPolyDataCompanion.scala

```diff
@@ -1,26 +1,35 @@
 package com.avsystem.commons
 package mongo.typed
 
-import com.avsystem.commons.annotation.explicitGenerics
+import com.avsystem.commons.annotation.{bincompat, explicitGenerics}
 import com.avsystem.commons.meta.MacroInstances
 import com.avsystem.commons.mongo.BsonGenCodecs
 import com.avsystem.commons.serialization.{GenCodec, GenObjectCodec}
 
 import scala.annotation.compileTimeOnly
 
 trait MongoPolyAdtInstances[D[_]] {
-  // needed by MongoAdtFormat.materialize for generic type
-  protected final implicit def codecFromFormat[T: MongoFormat]: GenCodec[T] = MongoFormat[T].codec
```

Comment:
I'm not familiarized well with this type structure, so here might go stupid question - the issue was that this was `protected` and it conflicted with `GenCodec` from the supplied `MacroInstances`?

Reply:
It goes a bit deeper than that.

What happens here is:
* The `MongoPolyAdtInstances.format` method is implemented by the macro `MongoAdtFormat.materialize` (it happens via the machinery of `MacroInstances` - read its scaladoc if you want to know more)
* Materialization of `MongoAdtFormat[T]` requires an implicit `GenObjectCodec[T]` instance (as defined by `MongoAdtFormat` itself)
* This `GenObjectCodec[T]` instance can be easily provided as `MongoPolyAdtInstances.codec[T](MongoFormat[T].codec)` but there is no implicit that does this
* Therefore, I added the `codecFromFormat` implicit that does it (in conjunction with `codec` method from the data type's companion)
* Unfortunately, `codecFromFormat` started conflicting with other implicits injected via `MacroInstances` machinery
* Therefore I just added the required implicit `GenCodec[T]` to the `format` method - which has more limited scope than the previously over-generic `codecFromFormat` method.

Yes, this isn't pretty - type-parameterized data types can be tricky with typeclass instance materialization. Sometimes it may be best to avoid them completely. Fortunately, they are easily avoidable in DB-related code if things turn out to be too complicated (at the price of some additional boilerplate).

Reply:
Thanks for the comprehensive explanation!